lxd/storage/drivers/generic_vfs.go

```diff
@@ -1194,3 +1195,144 @@ func genericVFSListVolumes(d Driver) ([]Volume, error) {
+		return err
+	}
+
+	revert.Add(func() { _ = cleanup() })
+
+	switch vol.contentType {
+	case ContentTypeFS:
+		mountPath := vol.MountPath()
+		if !filesystem.IsMountPoint(mountPath) {
+			err = vol.EnsureMountPath()
+			if err != nil {
+				return err
+			}
+
+			fsType := vol.ConfigBlockFilesystem()
+
+			if vol.mountFilesystemProbe {
+				fsType, err = fsProbe(volDevPath)
+				if err != nil {
+					return fmt.Errorf("Failed probing filesystem: %w", err)
+				}
+			}
+
+			mountFlags, mountOptions := filesystem.ResolveMountOptions(strings.Split(vol.ConfigBlockMountOptions(), ","))
+			err = TryMount(context.TODO(), volDevPath, mountPath, fsType, mountFlags, mountOptions)
+			if err != nil {
+				return err
+			}
+
+			d.Logger().Debug("Mounted volume", logger.Ctx{"driver": d.Info().Name, "volName": vol.name, "dev": volDevPath, "path": mountPath, "options": mountOptions})
+		}
+
+	case ContentTypeBlock:
+		// For VMs, mount the filesystem volume.
+		if vol.IsVMBlock() {
+			fsVol := vol.NewVMBlockFilesystemVolume()
+			err := d.MountVolume(fsVol, op)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	vol.MountRefCountIncrement() // From here on it is up to caller to call UnmountVolume() when done.
+	revert.Success()
+	return nil
+}
+
+func unmountVolume(d Driver, vol Volume, keepBlockDev bool, getDevicePath func(Volume, bool) (string, func() error, error), op *operations.Operation) (bool, error) {
+	unlock, err := vol.MountLock()
+	if err != nil {
+		return false, err
+	}
+
+	defer unlock()
+
+	ourUnmount := false
+	mountPath := vol.MountPath()
+	refCount := vol.MountRefCountDecrement()
+
+	// Attempt to unmount the volume.
+	if vol.contentType == ContentTypeFS && filesystem.IsMountPoint(mountPath) {
+		if refCount > 0 {
+			d.Logger().Debug("Skipping unmount as in use", logger.Ctx{"volName": vol.name, "refCount": refCount})
+			return false, ErrInUse
+		}
+
+		err := TryUnmount(mountPath, unix.MNT_DETACH)
+		if err != nil {
+			return false, err
+		}
+
+		d.Logger().Debug("Unmounted volume", logger.Ctx{"driver": d.Info().Name, "volName": vol.name, "path": mountPath, "keepBlockDev": keepBlockDev})
+
+		// Attempt to unmap.
+		if !keepBlockDev {
+			_, deactivate, err := getDevicePath(vol, false)
+			if err != nil {
+				return false, err
+			}
+
+			err = deactivate()
+			if err != nil {
+				return false, err
+			}
+		}
+
+		ourUnmount = true
+	} else if vol.contentType == ContentTypeBlock {
+		// For VMs, unmount the filesystem volume.
+		if vol.IsVMBlock() {
+			fsVol := vol.NewVMBlockFilesystemVolume()
+			ourUnmount, err = d.UnmountVolume(fsVol, false, op)
+			if err != nil {
+				return false, err
+			}
+		}
+
+		if !keepBlockDev {
+			// Check if device is currently mapped (but don't map if not).
+			devPath, _, _ := getDevicePath(vol, false)
+			if devPath != "" && shared.PathExists(devPath) {
+				if refCount > 0 {
+					d.Logger().Debug("Skipping unmount as in use", logger.Ctx{"volName": vol.name, "refCount": refCount})
```

Comment:
There's 2 calls to `d.Logger` in this function with similar context, so suggest creating a function specific logger context at the top of the function e.g.

```
l := d.Logger.().AddContext(logger.Ctx{"volName": vol.name})
```

Then use it, e.g. `l.Debug("Skipping unmount as in use", logger.Ctx{"refCount": refCount})`